File: app/src/main/java/com/astraedus/nudge/domain/engine/ScheduleEvaluator.kt
package com.astraedus.nudge.domain.engine

import com.astraedus.nudge.domain.model.ActiveRule
import java.util.Calendar
import javax.inject.Inject

/**
 * Determines whether a scheduled rule is currently active based on day-of-week and time-of-day.
 *
 * Day numbering: 1=Monday .. 7=Sunday (ISO 8601).
 * Time is expressed as minutes from midnight (0..1439).
 * Overnight schedules (end < start) are supported -- e.g. 23:00-06:00.
 */
class ScheduleEvaluator @Inject constructor() {

    /**
     * Returns true if the rule should be active right now.
     * If no schedule fields are set (all null), the rule is always active.
     */
    fun isActiveNow(rule: ActiveRule): Boolean {
        return isActiveAt(rule, Calendar.getInstance())
    }

    /**
     * Testable overload that accepts an explicit [Calendar] for the current moment.
     */
    fun isActiveAt(rule: ActiveRule, now: Calendar): Boolean {
        val days = rule.scheduleDays
        val startMinute = rule.scheduleStartMinute
        val endMinute = rule.scheduleEndMinute

        // No schedule configured -- always active
        if (days == null && startMinute == null && endMinute == null) return true

        // Convert Calendar.DAY_OF_WEEK (Sun=1..Sat=7) to ISO (Mon=1..Sun=7)
        val currentDayIso = calendarDayToIso(now.get(Calendar.DAY_OF_WEEK))

        // Check day-of-week constraint
        if (days != null && days.isNotEmpty() && currentDayIso !in days) return false

        // Check time-of-day constraint
        if (startMinute != null && endMinute != null) {
            val currentMinute = now.get(Calendar.HOUR_OF_DAY) * 60 + now.get(Calendar.MINUTE)
            return if (startMinute <= endMinute) {
                // Normal range: e.g. 09:00 - 17:00
                currentMinute in startMinute until endMinute
            } else {
                // Overnight range: e.g. 23:00 - 06:00
                currentMinute >= startMinute || currentMinute < endMinute
            }
        }

        // Only days set (no time range) -- active all day on those days
        return true
    }

    companion object {
        /**
         * Convert [Calendar.DAY_OF_WEEK] (Sunday=1 .. Saturday=7)
         * to ISO 8601 (Monday=1 .. Sunday=7).
         */
        fun calendarDayToIso(calendarDay: Int): Int = when (calendarDay) {
            Calendar.MONDAY -> 1
            Calendar.TUESDAY -> 2
            Calendar.WEDNESDAY -> 3
            Calendar.THURSDAY -> 4
            Calendar.FRIDAY -> 5
            Calendar.SATURDAY -> 6
            Calendar.SUNDAY -> 7
            else -> calendarDay
        }
    }
}

File: extension/src/core/scheduleEvaluator.ts
/**
 * Schedule evaluation. PURE.
 *
 * Direct port of Android's `domain/engine/ScheduleEvaluator.kt`, including its exact
 * boundary semantics — the ported Kotlin tests are mirrored in tests/core/scheduleEvaluator.test.ts.
 *
 * Day numbering: ISO 8601, 1=Monday .. 7=Sunday.
 * Time: minutes from local midnight, 0..1439.
 * Overnight spans (end < start, e.g. 23:00–06:00) are supported.
 */

export interface ScheduleWindow {
  /** ISO day numbers. null or empty = every day. */
  days: number[] | null;
  /** Minutes from local midnight. Both null = all day (on the chosen days). */
  startMinute: number | null;
  endMinute: number | null;
}

/** Convert a JS `Date.getDay()` (0=Sun..6=Sat) to ISO (1=Mon..7=Sun). */
export function isoDayOfWeek(date: Date): number {
  const day = date.getDay();
  return day === 0 ? 7 : day;
}

/** Minutes elapsed since local midnight for `date`. */
export function minutesSinceMidnight(date: Date): number {
  return date.getHours() * 60 + date.getMinutes();
}

/**
 * True when the schedule window covers `now` (local time).
 *
 * With no schedule fields set at all the window is "always" — that is what makes an
 * unscheduled rule permanently active.
 */
export function isScheduleActiveAt(window: ScheduleWindow, now: Date): boolean {
  const { days, startMinute, endMinute } = window;

  // Nothing configured -> always active.
  if ((days === null || days.length === 0) && startMinute === null && endMinute === null) {
    return true;
  }

  if (days !== null && days.length > 0 && !days.includes(isoDayOfWeek(now))) {
    return false;
  }

  if (startMinute !== null && endMinute !== null) {
    const current = minutesSinceMidnight(now);
    if (startMinute <= endMinute) {
      // Normal range, end EXCLUSIVE: 09:00–17:00 is active at 09:00, not at 17:00.
      // (start === end therefore yields an empty window — never active — matching Android.)
      return current >= startMinute && current < endMinute;
    }
    // Overnight range: 23:00–06:00.
    return current >= startMinute || current < endMinute;
  }

  // Only days set -> active all day on those days.
  return true;
}

/**
 * Milliseconds from `now` until the next local midnight.
 *
 * Computed by constructing the next calendar day at 00:00 LOCAL rather than adding a
 * fixed 24h, so a DST transition shortens/lengthens the interval correctly instead of
 * drifting the daily reset off midnight (ext-01 §4).
 */
export function msUntilNextLocalMidnight(now: Date): number {
  const next = new Date(now.getFullYear(), now.getMonth(), now.getDate() + 1, 0, 0, 0, 0);
  return next.getTime() - now.getTime();
}

/** Local calendar day key, `yyyy-mm-dd`. The key usage rollups are stored under. */
export function localDayKey(date: Date): string {
  const y = date.getFullYear();
  const m = String(date.getMonth() + 1).padStart(2, '0');
  const d = String(date.getDate()).padStart(2, '0');
  return `${y}-${m}-${d}`;
}
